# Pandas: Transforming Data

---

## Run The Following Code
Hit SHIFT + ENTER when your cusor is inside the cell of code.<br/>

## Importing and Checking the Data

> Importing pandas

In [60]:
import pandas as pd

> Read the data from CSV file into a DataFrame

In [61]:
hdb = 'data/0_raw_data/hdb_resale_sample_raw.csv'

df = pd.read_csv(hdb)

> viewing the contents of the dataset

The main dataset that we will work on is on HDB resale data.<br/>
See the description below for the fields given. <br/>
<br/>
<table>
    <tr><th>Field</th><th>Description</th></tr>
    <tr><td>Trans_YearMonth</td><td>year-month (YYYY-MM) for transaction</td></tr>
    <tr><td>town</td><td>town for the resale flat</td></tr>
    <tr><td>flat_type</td><td>flat type for the resale flat</td></tr>
    <tr><td>block</td><td>block for the resale flat</td></tr>
    <tr><td>street_name</td><td>street for the resale flat</td></tr>
    <tr><td>storey_range</td><td>storey range for the resale flat</td></tr>
    <tr><td>floor_area_sqm</td><td>floor area in square meter</td></tr>
    <tr><td>flat_model</td><td>model of resale flat</td></tr>
    <tr><td>lease_commence_date</td><td>year of lease commencement</td></tr>
    <tr><td>resale_price</td><td>price of the resale flat</td></tr>
</table>

In [62]:
df.shape #num of rows and columns 15000 rows, 10 columns

(15000, 10)

In [63]:
df.head() #first 5 rows of data

Tranc_YearMonth         town flat_type block       street_name storey_range  \
0         2015-01  ANG MOO KIO    3 ROOM   218  ANG MO KIO AVE 1     07 TO 09   
1         2015-01  ANG MOO KIO    4 ROOM   207  ANG MO KIO AVE 1     01 TO 03   
2         2015-02  ANG MOO KIO    3 ROOM   214  ANG MO KIO AVE 3     07 TO 09   
3         2015-04  ANG MOO KIO    3 ROOM   214  ANG MO KIO AVE 3     07 TO 09   
4         2015-04  ANG MOO KIO    3 ROOM   213  ANG MO KIO AVE 3     07 TO 09   

   floor_area_sqm      flat_model  lease_commence_date  resale_price  
0            67.0  New Generation                 1976      308000.0  
1            97.0  New Generation                 1976      475000.0  
2            67.0  New Generation                 1976      315000.0  
3            67.0  New Generation                 1976      285000.0  
4            67.0  New Generation                 1976      305000.0

## Sorting Rows in DataFrame

In [64]:
df.sort_values('resale_price') #default ascending

Tranc_YearMonth             town flat_type block         street_name  \
13185         2020-02        TOA PAYOH    3 ROOM    26      TOA PAYOH EAST   
4246          2019-10          GEYLANG    2 ROOM    68          CIRCUIT RD   
2273          2020-02      BUKIT MERAH    1 ROOM     7  TELOK BLANGAH CRES   
4279          2020-07          GEYLANG    2 ROOM    72          CIRCUIT RD   
4228          2019-07          GEYLANG    2 ROOM    71          CIRCUIT RD   
...               ...              ...       ...   ...                 ...   
6699          2015-03  KALLANG/WHAMPOA    3 ROOM    53          JLN MA'MOR   
1724          2020-08      BUKIT MERAH    4 ROOM    43        MOH GUAN TER   
6897          2021-04  KALLANG/WHAMPOA    3 ROOM    51         JLN BAHAGIA   
6746          2016-12  KALLANG/WHAMPOA    3 ROOM    57          JLN MA'MOR   
6806          2018-09  KALLANG/WHAMPOA    3 ROOM    41         JLN BAHAGIA   

      storey_range  floor_area_sqm     flat_model  lease_commence_date  \
13185     10 TO 12            67.0       Improved                 1971   
4246      04 TO 06            45.0       Standard                 1969   
2273      10 TO 12            31.0       Improved                 1975   
4279      10 TO 12            42.0       Standard                 1969   
4228      04 TO 06            42.0       Standard                 1969   
...            ...             ...            ...                  ...   
6699      01 TO 03           280.0        Terrace                 1972   
1724      04 TO 06           131.0  Adjoined flat                 1973   
6897      01 TO 03           178.0        Terrace                 1972   
6746      01 TO 03           259.0        Terrace                 1972   
6806      01 TO 03           237.0        Terrace                 1972   

       resale_price  
13185      140000.0  
4246       150000.0  
2273       157000.0  
4279       160000.0  
4228       160000.0  
...             ...  
6699      1060000.0  
1724      1088000.0  
6897      1140000.0  
6746      1150000.0  
6806      1185000.0  

[15000 rows x 10 columns]

In [65]:
df.sort_values('resale_price', ascending=False) #descending

Tranc_YearMonth             town flat_type block         street_name  \
6806          2018-09  KALLANG/WHAMPOA    3 ROOM    41         JLN BAHAGIA   
6746          2016-12  KALLANG/WHAMPOA    3 ROOM    57          JLN MA'MOR   
6897          2021-04  KALLANG/WHAMPOA    3 ROOM    51         JLN BAHAGIA   
1724          2020-08      BUKIT MERAH    4 ROOM    43        MOH GUAN TER   
6699          2015-03  KALLANG/WHAMPOA    3 ROOM    53          JLN MA'MOR   
...               ...              ...       ...   ...                 ...   
2276          2020-03      BUKIT MERAH    1 ROOM     7  TELOK BLANGAH CRES   
4633          2018-09          GEYLANG    2 ROOM    39          CIRCUIT RD   
2273          2020-02      BUKIT MERAH    1 ROOM     7  TELOK BLANGAH CRES   
4246          2019-10          GEYLANG    2 ROOM    68          CIRCUIT RD   
13185         2020-02        TOA PAYOH    3 ROOM    26      TOA PAYOH EAST   

      storey_range  floor_area_sqm     flat_model  lease_commence_date  \
6806      01 TO 03           237.0        Terrace                 1972   
6746      01 TO 03           259.0        Terrace                 1972   
6897      01 TO 03           178.0        Terrace                 1972   
1724      04 TO 06           131.0  Adjoined flat                 1973   
6699      01 TO 03           280.0        Terrace                 1972   
...            ...             ...            ...                  ...   
2276      01 TO 03            31.0       Improved                 1975   
4633      04 TO 06            47.0       Standard                 1971   
2273      10 TO 12            31.0       Improved                 1975   
4246      04 TO 06            45.0       Standard                 1969   
13185     10 TO 12            67.0       Improved                 1971   

       resale_price  
6806      1185000.0  
6746      1150000.0  
6897      1140000.0  
1724      1088000.0  
6699      1060000.0  
...             ...  
2276       160000.0  
4633       160000.0  
2273       157000.0  
4246       150000.0  
13185      140000.0  

[15000 rows x 10 columns]

In [66]:
df.sort_values(['lease_commence_date', 'resale_price'], ascending=False) #descending by 2 columns

Tranc_YearMonth         town flat_type block       street_name  \
11639         2019-08   QUEENSTOWN    5 ROOM     3       GHIM MOH RD   
2831          2018-09  BUKIT MERAH    5 ROOM    48  TELOK BLANGAH DR   
11541         2016-08   QUEENSTOWN    5 ROOM     2       GHIM MOH RD   
2946          2020-01  BUKIT MERAH    5 ROOM    48  TELOK BLANGAH DR   
2738          2017-05  BUKIT MERAH    5 ROOM    49  TELOK BLANGAH DR   
...               ...          ...       ...   ...               ...   
5400          2020-01  JURONG EAST    3 ROOM    37     TEBAN GDNS RD   
5398          2019-07  JURONG EAST    3 ROOM    39     TEBAN GDNS RD   
5397          2018-09  JURONG EAST    3 ROOM    37     TEBAN GDNS RD   
5401          2020-03  JURONG EAST    3 ROOM    38     TEBAN GDNS RD   
5393          2017-10  JURONG EAST    3 ROOM    37     TEBAN GDNS RD   

      storey_range  floor_area_sqm     flat_model  lease_commence_date  \
11639     13 TO 15           150.0  Adjoined flat                 1976   
2831      13 TO 15           123.0       Standard                 1976   
11541     04 TO 06           150.0  Adjoined flat                 1976   
2946      22 TO 24           123.0       Standard                 1976   
2738      22 TO 24           125.0       Standard                 1976   
...            ...             ...            ...                  ...   
5400      04 TO 06            67.0       Improved                 1966   
5398      01 TO 03            67.0       Improved                 1966   
5397      01 TO 03            67.0       Improved                 1966   
5401      04 TO 06            67.0       Improved                 1966   
5393      01 TO 03            67.0       Improved                 1966   

       resale_price  
11639      860000.0  
2831       830000.0  
11541      830000.0  
2946       825000.0  
2738       820000.0  
...             ...  
5400       220000.0  
5398       218000.0  
5397       215000.0  
5401       210000.0  
5393       200000.0  

[15000 rows x 10 columns]

In [67]:
df.sort_values(['lease_commence_date', 'resale_price'], ascending=[False,True]) #specify different orders by 2 columns

Tranc_YearMonth         town flat_type block       street_name  \
14881         2019-04    WOODLANDS    3 ROOM    15    MARSILING LANE   
6180          2019-08  JURONG WEST    3 ROOM   208       BOON LAY PL   
6147          2018-10  JURONG WEST    3 ROOM   209       BOON LAY PL   
6189          2019-09  JURONG WEST    3 ROOM   208       BOON LAY PL   
2975          2020-07  BUKIT MERAH    2 ROOM    45  TELOK BLANGAH DR   
...               ...          ...       ...   ...               ...   
5389          2015-03  JURONG EAST    3 ROOM    39     TEBAN GDNS RD   
5390          2015-08  JURONG EAST    3 ROOM    37     TEBAN GDNS RD   
5406          2013-07  JURONG EAST    3 ROOM    38     TEBAN GDNS RD   
5405          2013-04  JURONG EAST    3 ROOM    37     TEBAN GDNS RD   
5404          2013-01  JURONG EAST    3 ROOM    39     TEBAN GDNS RD   

      storey_range  floor_area_sqm flat_model  lease_commence_date  \
14881     01 TO 03            65.0   Improved                 1976   
6180      01 TO 03            65.0   Improved                 1976   
6147      04 TO 06            65.0   Improved                 1976   
6189      07 TO 09            65.0   Improved                 1976   
2975      01 TO 03            45.0   Improved                 1976   
...            ...             ...        ...                  ...   
5389      01 TO 03            67.0   Improved                 1966   
5390      04 TO 06            67.0   Improved                 1966   
5406      01 TO 03            67.0   Improved                 1966   
5405      04 TO 06            67.0   Improved                 1966   
5404      01 TO 03            67.0   Improved                 1966   

       resale_price  
14881      183500.0  
6180       185000.0  
6147       190000.0  
6189       190000.0  
2975       194000.0  
...             ...  
5389       260000.0  
5390       289000.0  
5406       310000.0  
5405       311000.0  
5404       315000.0  

[15000 rows x 10 columns]

## Handling Missing & Dirty Values

> Dealing with null values and dirty data

In [68]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Tranc_YearMonth      15000 non-null  object 
 1   town                 14429 non-null  object 
 2   flat_type            15000 non-null  object 
 3   block                15000 non-null  object 
 4   street_name          15000 non-null  object 
 5   storey_range         15000 non-null  object 
 6   floor_area_sqm       15000 non-null  float64
 7   flat_model           15000 non-null  object 
 8   lease_commence_date  15000 non-null  int64  
 9   resale_price         15000 non-null  float64
dtypes: float64(2), int64(1), object(7)
memory usage: 1.1+ MB


> remove null values

In [69]:
# this does not store the result
# we will need to assign the output to a variable or overwrite the current df
# df = df.dropna()
df.dropna()

Tranc_YearMonth         town flat_type block       street_name  \
0             2015-01  ANG MOO KIO    3 ROOM   218  ANG MO KIO AVE 1   
1             2015-01  ANG MOO KIO    4 ROOM   207  ANG MO KIO AVE 1   
2             2015-02  ANG MOO KIO    3 ROOM   214  ANG MO KIO AVE 3   
3             2015-04  ANG MOO KIO    3 ROOM   214  ANG MO KIO AVE 3   
4             2015-04  ANG MOO KIO    3 ROOM   213  ANG MO KIO AVE 3   
...               ...          ...       ...   ...               ...   
14995         2013-04    WOODLANDS    5 ROOM    11      MARSILING DR   
14996         2013-04    WOODLANDS    5 ROOM     9      MARSILING DR   
14997         2013-05    WOODLANDS    3 ROOM     6      MARSILING DR   
14998         2013-05    WOODLANDS    4 ROOM     7      MARSILING DR   
14999         2013-05    WOODLANDS    5 ROOM    10      MARSILING DR   

      storey_range  floor_area_sqm      flat_model  lease_commence_date  \
0         07 TO 09            67.0  New Generation                 1976   
1         01 TO 03            97.0  New Generation                 1976   
2         07 TO 09            67.0  New Generation                 1976   
3         07 TO 09            67.0  New Generation                 1976   
4         07 TO 09            67.0  New Generation                 1976   
...            ...             ...             ...                  ...   
14995     19 TO 21           123.0        Standard                 1976   
14996     19 TO 21           117.0        Standard                 1976   
14997     04 TO 06            65.0        Improved                 1976   
14998     04 TO 06            88.0        Improved                 1976   
14999     07 TO 09           123.0        Standard                 1976   

       resale_price  
0          308000.0  
1          475000.0  
2          315000.0  
3          285000.0  
4          305000.0  
...             ...  
14995      445000.0  
14996      510000.0  
14997      292000.0  
14998      360000.0  
14999      440000.0  

[14429 rows x 10 columns]

> update null values

In [70]:
# This replace the missing values (NaN) with the string 'BEDOK'
df['town'] = df['town'].fillna('BEDOK')

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Tranc_YearMonth      15000 non-null  object 
 1   town                 15000 non-null  object 
 2   flat_type            15000 non-null  object 
 3   block                15000 non-null  object 
 4   street_name          15000 non-null  object 
 5   storey_range         15000 non-null  object 
 6   floor_area_sqm       15000 non-null  float64
 7   flat_model           15000 non-null  object 
 8   lease_commence_date  15000 non-null  int64  
 9   resale_price         15000 non-null  float64
dtypes: float64(2), int64(1), object(7)
memory usage: 1.1+ MB


> replace values in a DataFrame's column

In [72]:
df['town'].unique()

array(['ANG MOO KIO', 'ANG MO KIO', 'BEDOK', 'BISHAN', 'BUKIT MERAH',
       'BUKIT TIMAH', 'CENTRAL AREA', 'GEYLANG', 'HOUGANG', 'JURONG EAST',
       'JURONG WEST', 'KALLANG/WHAMPOA', 'MARINE PARADE', 'QUEENSTOWN',
       'TOA PAYOH', 'WOODLANDS'], dtype=object)

In [73]:
# Remember to reassigned the revised values into existing column (or a new column)
df['town'] = df['town'].replace('ANG MOO KIO', 'ANG MO KIO')

In [74]:
df['town'].unique()

array(['ANG MO KIO', 'BEDOK', 'BISHAN', 'BUKIT MERAH', 'BUKIT TIMAH',
       'CENTRAL AREA', 'GEYLANG', 'HOUGANG', 'JURONG EAST', 'JURONG WEST',
       'KALLANG/WHAMPOA', 'MARINE PARADE', 'QUEENSTOWN', 'TOA PAYOH',
       'WOODLANDS'], dtype=object)

## Aggregating Numbers based on Category

> GroupBy

In [75]:
groupby_town = df.groupby('town')

In [76]:
groupby_town['resale_price'].mean() #mean of numerical values, you can try other Math functions!

town
ANG MO KIO         326890.235294
BEDOK              360959.922942
BISHAN             295249.210084
BUKIT MERAH        381027.219172
BUKIT TIMAH        550988.675159
CENTRAL AREA       382266.286996
GEYLANG            287563.004905
HOUGANG            300874.426667
JURONG EAST        245500.000000
JURONG WEST        265373.744797
KALLANG/WHAMPOA    369058.760257
MARINE PARADE      507970.847682
QUEENSTOWN         380540.040733
TOA PAYOH          318940.731699
WOODLANDS          273703.839506
Name: resale_price, dtype: float64

In [77]:
groupby_town_flattype = df.groupby(['town','flat_type'])

In [78]:
groupby_town_flattype['resale_price'].mean()

town             flat_type
ANG MO KIO       3 ROOM       309624.379562
                 4 ROOM       398569.696970
BEDOK            3 ROOM       299674.814070
                 4 ROOM       378142.461538
                 5 ROOM       639348.487805
BISHAN           3 ROOM       278461.940594
                 4 ROOM       376375.000000
                 5 ROOM       494000.000000
BUKIT MERAH      1 ROOM       206901.696970
                 2 ROOM       256891.258216
                 3 ROOM       362486.258861
                 4 ROOM       465457.716088
                 5 ROOM       671653.861635
BUKIT TIMAH      3 ROOM       416806.256410
                 4 ROOM       559840.291667
                 5 ROOM       874903.225806
CENTRAL AREA     2 ROOM       266767.681818
                 3 ROOM       398194.405063
                 4 ROOM       504423.238095
GEYLANG          2 ROOM       225962.870588
                 3 ROOM       274412.751502
                 4 ROOM       465122.857143
     

> Pivot Table

In [79]:
df.pivot_table(values='resale_price', index='town', columns='flat_type', aggfunc='mean')

flat_type              1 ROOM         2 ROOM         3 ROOM         4 ROOM  \
town                                                                         
ANG MO KIO                NaN            NaN  309624.379562  398569.696970   
BEDOK                     NaN            NaN  299674.814070  378142.461538   
BISHAN                    NaN            NaN  278461.940594  376375.000000   
BUKIT MERAH      206901.69697  256891.258216  362486.258861  465457.716088   
BUKIT TIMAH               NaN            NaN  416806.256410  559840.291667   
CENTRAL AREA              NaN  266767.681818  398194.405063  504423.238095   
GEYLANG                   NaN  225962.870588  274412.751502  465122.857143   
HOUGANG                   NaN            NaN  258502.705882  348261.411765   
JURONG EAST               NaN            NaN  245500.000000            NaN   
JURONG WEST               NaN  210800.000000  254670.994097  302000.000000   
KALLANG/WHAMPOA           NaN  250901.375000  339375.547464  442574.228374   
MARINE PARADE             NaN  245857.142857  399567.667257  521356.759717   
QUEENSTOWN                NaN  244910.376471  354952.585253  544913.954918   
TOA PAYOH                 NaN  229655.162162  299230.152174  401168.859694   
WOODLANDS                 NaN            NaN  254830.024823  279292.986301   

flat_type               5 ROOM  
town                            
ANG MO KIO                 NaN  
BEDOK            639348.487805  
BISHAN           494000.000000  
BUKIT MERAH      671653.861635  
BUKIT TIMAH      874903.225806  
CENTRAL AREA               NaN  
GEYLANG          676912.000000  
HOUGANG          494500.000000  
JURONG EAST                NaN  
JURONG WEST      420398.283333  
KALLANG/WHAMPOA  590965.225806  
MARINE PARADE    801507.381188  
QUEENSTOWN       771895.862069  
TOA PAYOH        547929.495327  
WOODLANDS        371992.000000

## Processing Data with Customized Logics

### Involving Values from One Column

In [80]:
#Splitting year month
def get_year(row): 
    yearmonth_string = row['Tranc_YearMonth']
    year_string = yearmonth_string.split('-')[0]
    year_string = int(year_string)
    return year_string

In [81]:
# The apply() method will produce output as a Pandas Series object
# Which is similar a column of values
# So if we don't assign/store the output, it will just display on Jupyter Notebook
# !!!! Always remember to use axis=1
df.apply(get_year, axis=1) 

0        2015
1        2015
2        2015
3        2015
4        2015
         ... 
14995    2013
14996    2013
14997    2013
14998    2013
14999    2013
Length: 15000, dtype: int64

In [82]:
# Remember to assign the output to a new column (or overwriting the existing column if applicabble)
df['Tranc_Year'] = df.apply(get_year, axis=1) 

### Involving Values from Multple Columns

In [83]:
#Function for row data
def get_hdb_type_model(row):
    type_model = row['flat_type'] + ' ' + row['flat_model']
    return type_model

In [84]:
df.apply(get_hdb_type_model, axis=1) #applying function to all rows

0        3 ROOM New Generation
1        4 ROOM New Generation
2        3 ROOM New Generation
3        3 ROOM New Generation
4        3 ROOM New Generation
                 ...          
14995          5 ROOM Standard
14996          5 ROOM Standard
14997          3 ROOM Improved
14998          4 ROOM Improved
14999          5 ROOM Standard
Length: 15000, dtype: object

In [85]:
df['flat_type_model'] = df.apply(get_hdb_type_model, axis=1) #assign as new column

## Exporting the Processed DataFrame

In [86]:
output_filepath = 'data/1_Processed_Data/hdb_resale_sample_processed.csv'

# set "index=False" to exclude the index being exported as part of the CSV
df.to_csv(output_filepath, index=False)

# [OPTIONAL]: Extra Learning

## Quickly Passing Data between Pandas and Excel

In [87]:
# After running this cell, head over a new Excel's Spreahsheet and paste (Ctrl + V)
df_temp = df.head(100)
df_temp.to_clipboard(excel=True, index=False)

In [88]:
# The reverse is also possible
# Select the cells (i.e. ranges) on an Excel's Spreahsheet
# Then run the this cell
df_temp = pd.read_clipboard()

## Exporting DataFrame as a Static Webpage (HTML File)

In [89]:
df_temp = df.head(100)
df_temp.to_html('data/export_dataframe_as_webpage.html')


<br>

---

---

# !! Your Turn !!

## Replace the code marked as <..> or fill the empty cell COMPLETELY with your own code
Don't constrained by the <..>, it's just a guide. Feel free to insert more lines of code.
Or using less lines of code is also fine.

<br>

---

# Question 1 
A) Read hdb_resale_full_raw.csv into pandas and named it <b>df</b> <br/>

In [139]:

df = pd.read_csv('data/0_raw_data/hdb_resale_full_raw.csv')

<br>
B) Find out which column(s) has records with missing values

💡 Hint: use info() or any other methods

In [140]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 185968 entries, 0 to 185967
Data columns (total 10 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   Tranc_YearMonth      184086 non-null  object 
 1   town                 185968 non-null  object 
 2   flat_type            185945 non-null  object 
 3   block                185968 non-null  object 
 4   street_name          185968 non-null  object 
 5   storey_range         185968 non-null  object 
 6   floor_area_sqm       185968 non-null  float64
 7   flat_model           185968 non-null  object 
 8   lease_commence_date  185968 non-null  int64  
 9   resale_price         185968 non-null  float64
dtypes: float64(2), int64(1), object(7)
memory usage: 14.2+ MB


In [92]:
# Method 2: 
df.isnull().sum()

Tranc_YearMonth        1882
town                      0
flat_type                23
block                     0
street_name               0
storey_range              0
floor_area_sqm            0
flat_model                0
lease_commence_date       0
resale_price              0
dtype: int64

<br>

---

# Question 2

<br>

A) Let's examine the **Tranc_YearMonth column** by listing the all unique values in the dataset <br/>

In [93]:
df['Tranc_YearMonth'].unique()

array(['2012-03', '2012-04', '2012-05', '2012-06', '2012-07', '2012-08',
       '2012-09', '2012-10', '2012-11', '2012-12', '2013-01', '2013-02',
       '2013-03', '2013-04', '2013-05', '2013-06', '2013-07', '2013-08',
       '2013-09', '2013-10', '2013-11', '2013-12', '2014-01', '2014-02',
       '2014-03', '2014-04', '2014-05', '2014-06', '2014-07', '2014-08',
       '2014-09', '2014-10', '2014-11', '2014-12', '2015-01', '2015-02',
       '2015-03', '2015-04', '2015-05', '2015-06', '2015-07', '2015-08',
       '2015-09', '2015-10', '2015-11', '2015-12', '2016-01', '2016-02',
       '2016-03', '2016-04', '2016-05', '2016-06', '2016-07', nan,
       '2016-09', '2016-10', '2016-11', '2016-12', '2017-01', '2017-02',
       '2017-03', '2017-04', '2017-05', '2017-06', '2017-07', '2017-08',
       '2017-09', '2017-10', '2017-11', '2017-12', '2018-01', '2018-02',
       '2018-03', '2018-04', '2018-05', '2018-06', '2018-07', '2018-08',
       '2018-09', '2018-10', '2018-11', '2018-12', '2019-

<br>

B) You are being informed that the records are ordered from the **earliest** to the **latest** transactions.\
Use this information to fill the missing values in the **Tranc_YearMonth** column. \
⚠️ Note: Remember to assign the values back to the column

In [141]:
# This replace the missing date "nan" with the string '2016-08'
df['Tranc_YearMonth'] = df['Tranc_YearMonth'].fillna('2016-08')

# Confirm that the nan value in the column is replaced.
df['Tranc_YearMonth'].isnull().sum()

0

---

<br>

B) You are also being informed there are typos for same values in **flat_type** column due to some errors.\
There are "3 ROOM" and "3 ROOMS". **Correct the typos**.

💡 # Hint: Use .replace() or any other methods

⚠️ Note: Ignore the missing values in this column for now.

In [95]:
df['flat_type'].unique()

array(['2 ROOM', '3 ROOM', '4 ROOM', '5 ROOM', 'EXECUTIVE',
       'MULTI-GENERATION', '1 ROOM', '3 ROOMS', nan], dtype=object)

In [142]:
df['flat_type']= df['flat_type'].replace('3 ROOMS', '3 ROOM')

In [97]:
df['flat_type'].unique()

array(['2 ROOM', '3 ROOM', '4 ROOM', '5 ROOM', 'EXECUTIVE',
       'MULTI-GENERATION', '1 ROOM', nan], dtype=object)

<br>

---

# Question 3

A) Retrieve all the transactions for **4-room flats** from **Clementi** and\
the records must be in the order which the **latest transactions are at the top of the DataFrame**.\
Note: There is no need to store the final output. Just display in Jupyter Notebook.

In [98]:
df_temp = df[(df['town'] == 'CLEMENTI') & (df['flat_type'] =='4 ROOM')]
df_temp.sort_values('Tranc_YearMonth', ascending=False )

Tranc_YearMonth      town flat_type block         street_name  \
184631         2021-04  CLEMENTI    4 ROOM  312C      CLEMENTI AVE 4   
184614         2021-04  CLEMENTI    4 ROOM  311A      CLEMENTI AVE 4   
184605         2021-04  CLEMENTI    4 ROOM   425      CLEMENTI AVE 1   
184606         2021-04  CLEMENTI    4 ROOM   417      CLEMENTI AVE 1   
184607         2021-04  CLEMENTI    4 ROOM   358      CLEMENTI AVE 2   
...                ...       ...       ...   ...                 ...   
758            2012-03  CLEMENTI    4 ROOM   411   C'WEALTH AVE WEST   
760            2012-03  CLEMENTI    4 ROOM   113      CLEMENTI ST 13   
761            2012-03  CLEMENTI    4 ROOM   454      CLEMENTI AVE 3   
764            2012-03  CLEMENTI    4 ROOM   461      CLEMENTI AVE 3   
744            2012-03  CLEMENTI    4 ROOM   709  CLEMENTI WEST ST 2   

       storey_range  floor_area_sqm      flat_model  lease_commence_date  \
184631     37 TO 39            93.0         Model A                 2017   
184614     22 TO 24            80.0            DBSS                 2014   
184605     13 TO 15           100.0         Model A                 2013   
184606     04 TO 06           100.0         Model A                 2013   
184607     04 TO 06            92.0  New Generation                 1978   
...             ...             ...             ...                  ...   
758        11 TO 15            98.0  New Generation                 1978   
760        11 TO 15           107.0         Model A                 1984   
761        06 TO 10            86.0         Model A                 2002   
764        06 TO 10            90.0         Model A                 2006   
744        01 TO 05            91.0  New Generation                 1980   

        resale_price  
184631      832888.0  
184614      790000.0  
184605      695000.0  
184606      690000.0  
184607      465000.0  
...              ...  
758         555000.0  
760         600000.0  
761         650000.0  
764         677000.0  
744         435000.0  

[1422 rows x 10 columns]

<br>

---

B) Sort the records in the **df**, which the records are ordered by:
1. Lease Commence Year (Newest to Oldest)
2. Transaction Year (Newest to Oldest)
3. Town (In alphabetical order)
4. Flat Type (In alphabetical order)
5. Resale Price (Lowest to Highest Price)

Note: There is no need to store the final output. Just display in Jupyter Notebook.

Your DataFrame should look something like this.\
![](data/Q3.jpg)

In [99]:
df.sort_values(
  ['lease_commence_date', 'Tranc_YearMonth', 'town', 'flat_type', 'resale_price'], 
  ascending=[False, False, True, True, True])

Tranc_YearMonth         town flat_type block    street_name  \
182024         2021-03  BUKIT MERAH    3 ROOM   96B   HENDERSON RD   
182023         2021-03  BUKIT MERAH    3 ROOM   95A   HENDERSON RD   
182028         2021-03  BUKIT MERAH    4 ROOM   96A   HENDERSON RD   
182027         2021-03  BUKIT MERAH    4 ROOM   96A   HENDERSON RD   
179832         2021-02  BUKIT MERAH    3 ROOM   95B   HENDERSON RD   
...                ...          ...       ...   ...            ...   
63013          2015-08  JURONG EAST    3 ROOM    37  TEBAN GDNS RD   
55171          2015-03  JURONG EAST    3 ROOM    39  TEBAN GDNS RD   
29021          2013-07  JURONG EAST    3 ROOM    38  TEBAN GDNS RD   
24484          2013-04  JURONG EAST    3 ROOM    37  TEBAN GDNS RD   
20611          2013-01  JURONG EAST    3 ROOM    39  TEBAN GDNS RD   

       storey_range  floor_area_sqm flat_model  lease_commence_date  \
182024     04 TO 06            68.0    Model A                 2019   
182023     40 TO 42            68.0    Model A                 2019   
182028     07 TO 09            93.0    Model A                 2019   
182027     25 TO 27            93.0    Model A                 2019   
179832     22 TO 24            68.0    Model A                 2019   
...             ...             ...        ...                  ...   
63013      04 TO 06            67.0   Improved                 1966   
55171      01 TO 03            67.0   Improved                 1966   
29021      01 TO 03            67.0   Improved                 1966   
24484      04 TO 06            67.0   Improved                 1966   
20611      01 TO 03            67.0   Improved                 1966   

        resale_price  
182024      620000.0  
182023      680000.0  
182028      820000.0  
182027      868000.0  
179832      660000.0  
...              ...  
63013       289000.0  
55171       260000.0  
29021       310000.0  
24484       311000.0  
20611       315000.0  

[185968 rows x 10 columns]

<br>

---

# Question 4
A) Find out the highest resale price value for 4-ROOM flat across all the towns.\
To simplify this, assume the calculation is
- based on the resale price only
- regardless of the age of the flats

Note: Just display the resale price in Jupyter Notebook

In [100]:
df[df['flat_type']=="4 ROOM"].resale_price.max()

1186888.0

<br>

B) Display the record for the transactio from A) above

The output should look something like this:\
![](data/Q4.jpg)

In [101]:
highest_resale_price = 1186888.0

df[df['resale_price']==highest_resale_price]

Tranc_YearMonth          town flat_type block    street_name  \
160493         2020-06  CENTRAL AREA    4 ROOM    1B  CANTONMENT RD   

       storey_range  floor_area_sqm flat_model  lease_commence_date  \
160493     40 TO 42            93.0    Type S1                 2011   

        resale_price  
160493     1186888.0

<br>
   
C) Find out the what is the highest price for the different **Flat Types**

💡Hint: use .groupby() method\
Note: Just display the resale price in Jupyter Notebook

In [102]:
groupby_flattype = df.groupby('flat_type')

In [103]:
groupby_flattype['resale_price'].max()

flat_type
1 ROOM               260000.0
2 ROOM               510000.0
3 ROOM              1185000.0
4 ROOM              1186888.0
5 ROOM              1258000.0
EXECUTIVE           1210000.0
MULTI-GENERATION     990000.0
Name: resale_price, dtype: float64

<br>
D) For each of the towns, find out what is the highest resale price for the different Flat Types.

Store the result in new DataFrame, called **df_highest_price**.

💡Hint: You might need .reset_index() to convert the result into a typical DataFrame

In [104]:
groupby_town_flattype = df.groupby(['town', 'flat_type'])

df_highest_price =groupby_town_flattype['resale_price'].max()

df_highest_price.reset_index() 

town         flat_type  resale_price
0    ANG MO KIO            2 ROOM      315000.0
1    ANG MO KIO            3 ROOM      570000.0
2    ANG MO KIO            4 ROOM      782000.0
3    ANG MO KIO            5 ROOM     1008888.0
4    ANG MO KIO         EXECUTIVE     1030000.0
..          ...               ...           ...
123      YISHUN            3 ROOM      460000.0
124      YISHUN            4 ROOM      625000.0
125      YISHUN            5 ROOM      770000.0
126      YISHUN         EXECUTIVE      818000.0
127      YISHUN  MULTI-GENERATION      868000.0

[128 rows x 3 columns]

In [105]:
# No need to modify this cell. Just execute.
# if above codes are correct, this should return "True"
# The idea is that if the groupby result is in DataFrame, then we can use the data easily
type(df_highest_price) == pd.core.frame.DataFrame

False

In [106]:
# No need to modify this cell. Just execute.
# Display the result
df_highest_price

town        flat_type       
ANG MO KIO  2 ROOM               315000.0
            3 ROOM               570000.0
            4 ROOM               782000.0
            5 ROOM              1008888.0
            EXECUTIVE           1030000.0
                                  ...    
YISHUN      3 ROOM               460000.0
            4 ROOM               625000.0
            5 ROOM               770000.0
            EXECUTIVE            818000.0
            MULTI-GENERATION     868000.0
Name: resale_price, Length: 128, dtype: float64

---

# Question 5

A) Find the **number of transactions** which **resale price** is **less than $500,000** for each town. \
**<u>Optional</u>**: Display the result, with the **town with most number of transaction first row** and the **town with least number of transaction as the last row**.
<br/>

💡 Hint: There is an aggregation method for the groupby object, called **.count()&** and **.size()**.\
You may look up for more information on how use these two methods.

In [107]:
df_temp = df[df['resale_price'] < 500000].groupby('town')
print(df_temp.size().sort_values(ascending=False))

town
WOODLANDS          12163
JURONG WEST        11399
YISHUN             10939
SENGKANG            9618
BEDOK               8395
TAMPINES            8126
HOUGANG             6884
ANG MO KIO          6801
CHOA CHU KANG       6578
PUNGGOL             6459
BUKIT BATOK         5535
BUKIT PANJANG       5290
SEMBAWANG           4032
TOA PAYOH           3647
GEYLANG             3440
PASIR RIS           3342
KALLANG/WHAMPOA     3094
JURONG EAST         3055
BUKIT MERAH         2988
CLEMENTI            2915
QUEENSTOWN          2479
SERANGOON           2297
BISHAN               897
CENTRAL AREA         750
MARINE PARADE        670
BUKIT TIMAH           89
dtype: int64


B) Based on the transactions where the  **resale price** is **less than $500,000**,\
generate a pivot table to show the **median** **resale price** for the different **flat types** from each of the towns.

The output looks something like this:\
![](data/Q5.jpg)

In [108]:
df_temp = df[df['resale_price'] < 500000]
df_temp.pivot_table(values='resale_price', index='town', columns='flat_type', aggfunc='median')

flat_type          1 ROOM    2 ROOM    3 ROOM    4 ROOM    5 ROOM  EXECUTIVE
town                                                                        
ANG MO KIO            NaN  245000.0  320000.0  435000.0  474000.0        NaN
BEDOK                 NaN  240000.0  310000.0  408000.0  467000.0        NaN
BISHAN                NaN       NaN  375000.0  468000.0  473000.0        NaN
BUKIT BATOK           NaN       NaN  285000.0  395000.0  460000.0   490000.0
BUKIT MERAH      209000.0  260000.0  350000.0  459000.0       NaN        NaN
BUKIT PANJANG         NaN  236900.0  305000.0  390000.0  445000.0   460000.0
BUKIT TIMAH           NaN       NaN  415000.0  489000.0       NaN        NaN
CENTRAL AREA          NaN  275000.0  415000.0  465000.0       NaN        NaN
CHOA CHU KANG         NaN  235000.0  298000.0  365000.0  420000.0   470000.0
CLEMENTI              NaN  300000.0  339000.0  441000.0  480000.0        NaN
GEYLANG               NaN  228000.0  297000.0  408000.0  475000.0        NaN
HOUGANG               NaN  235000.0  308000.0  401194.0  460000.0   465000.0
JURONG EAST           NaN  240000.0  310000.0  408000.0  450000.0   495000.0
JURONG WEST           NaN  230000.0  288000.0  388000.0  445000.0   470000.0
KALLANG/WHAMPOA       NaN  255000.0  330000.0  450000.0  482000.0        NaN
MARINE PARADE         NaN  240000.0  399500.0  470000.0       NaN        NaN
PASIR RIS             NaN  260000.0  327000.0  420000.0  470000.0   484000.0
PUNGGOL               NaN  255000.0  360000.0  440000.0  438000.0   485000.0
QUEENSTOWN            NaN  256444.0  348000.0  450000.0  430000.0        NaN
SEMBAWANG             NaN  225000.0  298000.0  366000.0  420000.0   470000.0
SENGKANG              NaN  255000.0  350000.0  418000.0  445000.0   480000.0
SERANGOON             NaN  251500.0  330000.0  420000.0  475000.0        NaN
TAMPINES              NaN  265000.0  340000.0  430000.0  480000.0        NaN
TOA PAYOH             NaN  233000.0  308000.0  410000.0  480000.0        NaN
WOODLANDS             NaN  230000.0  285000.0  363000.0  428000.0   480000.0
YISHUN                NaN  225000.0  295000.0  370000.0  455000.0   485000.0

---

# Question 6

<br>

A) Create a new column **storey_midpoint** to store the middle value of the **storey_range**.\
For example, if the **storey_range** is '05 to 07", it should will be 6 (as an Integer).


💡Hint: use apply() method and you will need to define a function to be used with the apply() method.


In [109]:
def get_mid_storey(row):
    top = int(row[-2:])
    bottom = int(row[0:2])
    mid = (top + bottom)/2
    return int(mid)

In [110]:
df['storey_midpoint'] = df['storey_range'].apply(get_mid_storey)

In [111]:
df.head()

Tranc_YearMonth        town flat_type block        street_name storey_range  \
0         2012-03  ANG MO KIO    2 ROOM   172   ANG MO KIO AVE 4     06 TO 10   
1         2012-03  ANG MO KIO    2 ROOM   510   ANG MO KIO AVE 8     01 TO 05   
2         2012-03  ANG MO KIO    3 ROOM   610   ANG MO KIO AVE 4     06 TO 10   
3         2012-03  ANG MO KIO    3 ROOM   474  ANG MO KIO AVE 10     01 TO 05   
4         2012-03  ANG MO KIO    3 ROOM   604   ANG MO KIO AVE 5     06 TO 10   

   floor_area_sqm      flat_model  lease_commence_date  resale_price  \
0            45.0        Improved                 1986      250000.0   
1            44.0        Improved                 1980      265000.0   
2            68.0  New Generation                 1980      315000.0   
3            67.0  New Generation                 1984      320000.0   
4            67.0  New Generation                 1980      321000.0   

   storey_midpoint  
0                8  
1                3  
2                8  
3                3  
4                8

<br>

B) Create a new column **address** to that combine the **block number** and the **street name**.\
For example: the address for the first row is "172, ANG MO KIO AVE 4".


In [112]:
def get_lower_letters(row):
    address = row['block'] + ', ' + row['street_name']
    return address


In [113]:
df['address'] = df.apply(get_lower_letters, axis=1)

In [114]:
df.head()

Tranc_YearMonth        town flat_type block        street_name storey_range  \
0         2012-03  ANG MO KIO    2 ROOM   172   ANG MO KIO AVE 4     06 TO 10   
1         2012-03  ANG MO KIO    2 ROOM   510   ANG MO KIO AVE 8     01 TO 05   
2         2012-03  ANG MO KIO    3 ROOM   610   ANG MO KIO AVE 4     06 TO 10   
3         2012-03  ANG MO KIO    3 ROOM   474  ANG MO KIO AVE 10     01 TO 05   
4         2012-03  ANG MO KIO    3 ROOM   604   ANG MO KIO AVE 5     06 TO 10   

   floor_area_sqm      flat_model  lease_commence_date  resale_price  \
0            45.0        Improved                 1986      250000.0   
1            44.0        Improved                 1980      265000.0   
2            68.0  New Generation                 1980      315000.0   
3            67.0  New Generation                 1984      320000.0   
4            67.0  New Generation                 1980      321000.0   

   storey_midpoint                 address  
0                8   172, ANG MO KIO AVE 4  
1                3   510, ANG MO KIO AVE 8  
2                8   610, ANG MO KIO AVE 4  
3                3  474, ANG MO KIO AVE 10  
4                8   604, ANG MO KIO AVE 5

<br>

C) There are some missing values in the **flat_type** model.\
After confirming with the domain/business users, there are only 2 possibilities:

- if the floor_area_sqm is **at least 110 squared meters** then it's 5-room flat
- **otherwise** it's 4-room flat

you requested to use the following logics to fill up the missing values:\
hint: 
- you can use the .apply() method or any other approaches. 
- you will also need to check if the value is missing

Note: This is simplified logic solely for practice purpose

In [56]:
def fill_missing_flat_type(row):
    value_to_return = '' #empty string
    
    # Checking if the value in this column is missing
    if pd.isna(row['flat_type']):
        if row['floor_area_sqm'] >= 110:
            value_to_return = '5 ROOM'
        else:
            value_to_return = '4 ROOM'
    else:
        row['flat_type']       # return the original values if not null
    return value_to_return

In [57]:
df['flat_type'] = df.apply(fill_missing_flat_type, axis=1)

In [115]:
# Confirm nan values in 'flat type' column are replaced.
df['flat_type'].isna().sum()

23

In [116]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 185968 entries, 0 to 185967
Data columns (total 12 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   Tranc_YearMonth      185968 non-null  object 
 1   town                 185968 non-null  object 
 2   flat_type            185945 non-null  object 
 3   block                185968 non-null  object 
 4   street_name          185968 non-null  object 
 5   storey_range         185968 non-null  object 
 6   floor_area_sqm       185968 non-null  float64
 7   flat_model           185968 non-null  object 
 8   lease_commence_date  185968 non-null  int64  
 9   resale_price         185968 non-null  float64
 10  storey_midpoint      185968 non-null  int64  
 11  address              185968 non-null  object 
dtypes: float64(2), int64(2), object(8)
memory usage: 17.0+ MB


In [143]:
# Alternative method using boolean mask
# Speed to replace is faster than apply method
 
# mask1 = (df['flat_type'].isna()) & (df['floor_area_sqm']>=110)
# mask2 = (df['flat_type'].isna()) & (df['floor_area_sqm']<110)

# df['flat_type'][mask1] ='5 ROOM'
# df['flat_type'][mask2] ='4 ROOM'


<ipython-input-144-027df2000941>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['flat_type'][mask1] ='5 ROOM'
<ipython-input-144-027df2000941>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['flat_type'][mask2] ='4 ROOM'


In [145]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 185968 entries, 0 to 185967
Data columns (total 10 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   Tranc_YearMonth      185968 non-null  object 
 1   town                 185968 non-null  object 
 2   flat_type            185968 non-null  object 
 3   block                185968 non-null  object 
 4   street_name          185968 non-null  object 
 5   storey_range         185968 non-null  object 
 6   floor_area_sqm       185968 non-null  float64
 7   flat_model           185968 non-null  object 
 8   lease_commence_date  185968 non-null  int64  
 9   resale_price         185968 non-null  float64
dtypes: float64(2), int64(1), object(7)
memory usage: 14.2+ MB


# Exporting the Processed Data
Save the processed df in **Processed Data folder** as <b>hdb_resale_full_processed.csv</b> <br/>

In [146]:
output_filepath = 'data/1_Processed_Data/hdb_resale_full_processed.csv'

# set "index=False" to exclude the index being exported as part of the CSV
df.to_csv(output_filepath, index=False)